### PR TITLE
Add bd3lm

### PR DESCRIPTION
**Pull request by rjagalamari**
## What

Adds BD3-LM (Block Discrete Denoising Diffusion Language Model) to `xlm-models`,
ported from the kuleshov-group reference implementation.

## Contribution type

- [x] Maintained model (`xlm-models/`)

## Labels

`model`, `enhancement`

## Changes

- New `xlm-models/bd3lm/` package
- Configs for seq2seq (star easy/medium/hard) and unconditional pre-training (OWT)
- Model sizes `tiny`, `small`, `medium`
- Fine-tuning from the released BD3-LM checkpoints via `+pretrained=auto`
- Confidence-based decoding
## Testing

Verified by running the eval configs: 1.000 exact match on
star-small and star-medium, at batch sizes 1, 4 and 16. 
Fine-tuning from `kuleshov-group/bd3lm-owt-block_size4` loads and trains.

## Documentation

- [x] `xlm-models/bd3lm/README.md`
- - [x] `docs/models/bd3lm.md`

### CODE REVIEW

**Comment by rjagalamari** [comment]
> ## Review: pretrained checkpoint loading vs LLaDA/Dream, plus a generation smoke test
> Looked specifically at (1) whether loading `kuleshov-group/bd3lm-owt-*` is the same pathway LLaDA/Dream use, and (2) whether a loaded OWT checkpoint produces a sensible continuation of a made-up prompt.
> 
> ### 1. Loading path: this is a new pathway for the same objective
> LLaDA and Dream load released Hub weights through the **shared** xLM inference/train loader:
> 
> * experiment sets `hub.repo_id` (and the official tokenizer from that same repo)
> * `skip_init_weights: true`
> * `load_model_for_inference` / `lightning_train` → `download_model_weights` → `load_model_weights_into_model(..., strict=True)`
> * state-dict keys match 1:1 because those models wrap/vendor the HF architecture (`HF checkpoints load without key prefixing`)
> 
> BD3LM does **not** use that path. It adds a model-local loader:
> 
> * `+pretrained=auto` → top-level `pretrained_from: kuleshov-group/bd3lm-owt-block_size${block_size}`
> * `Bd3lmModel.__init__` calls `load_pretrained_bd3lm()`, which downloads the repo itself and strips a `backbone.` prefix
> * a second, parallel offline path (`convert_hf_checkpoint.py` + `model_only_checkpoint_path`) can feed the shared loader, but only after a converted local file exists
> 
> That prefix strip is a real difference versus LLaDA/Dream. The released checkpoint is an HF `BD3LM` module whose tensors are named `backbone.blocks.0.attn_qkv.weight`; this port flattened the DDiT to `blocks.0.attn_qkv.weight`. A `hub.repo_id=kuleshov-group/bd3lm-owt-block_size4` strict load therefore fails:
> 
> ```
> model keys:   blocks.0.adaLN_modulation.weight
> hub keys:     backbone.blocks.0.adaLN_modulation.weight
> ```
> 
> After `strip_backbone_prefix`, the **same** shared helper succeeds with `strict=True` (131/131 tensors, vocab 50258). So the conversion is justified, but it did not need a new Hydra group or an in-`__init__` download. Two ways to stay on the LLaDA/Dream path:
> 
> 1. Keep a `self.backbone` submodule so Hub keys already match, then use `hub.repo_id` + `strict_model_only_load: false` for the extra `sampling_eps_*` buffers.
> 2. Convert once (or remap inside `load_model_weights_into_model`) and load via `hub.repo_id` / `model_only_checkpoint_path` after construction, with `skip_init_weights`.
> 
> The in-`__init__` loader also skips `skip_init_weights` / `init_dtype`, reimplements Hub download, and will fight the shared loader if someone later sets `hub.repo_id` as well (init remaps, then train/generate tries to load the unconverted file).
> 
> ### 2. Tokenizer mismatch on the advertised OWT fine-tune path
> The released checkpoint is GPT-2 + **one** mask token: `vocab_embed.embedding` is `(50258, 768)`, mask id `50257`. That is what the kuleshov-group sampler does (`mask_index = tokenizer.vocab_size`, then `vocab_size += 1`).
> 
> `experiment=owt_bd3lm` uses `xlm.datamodule.GPT2TokenizerFast`, which **adds six** special tokens (`<|cls|>`, `<|bos|>`, `<|pad|>`, `<|unk|>`, `<|mask|>`, `<|sep|>`). Measured:
> 
> tokenizer	`len()`	mask id
> HF GPT-2 (official BD3LM)	50257 (+1 mask → 50258)	50257
> xLM `GPT2TokenizerFast`	**50263**	**50261** (`<
> `+pretrained=auto` on `owt_bd3lm` therefore skips the three vocab tensors (shape mismatch) and leaves embedding/lm-head **random**. LLaDA/Dream avoid this by loading the checkpoint’s own tokenizer (`load_auto_tokenizer` / `DreamTokenizer.from_pretrained` on the Hub repo). Skipping vocab tensors is correct for star-graph fine-tunes (vocab 20); it is not correct for the GPT-2 OWT checkpoint this flag is documented to load.
> 
> `Bd3lmPredictor` also has no `generate(prompts: List[str])`, so `job_type=demo` / `cli_demo.py` cannot drive it the way ILM/FlexMDM do. I called `_sample` directly.
> 
> ### 3. Inference smoke test (`kuleshov-group/bd3lm-owt-block_size4`)
> CPU, `block_size=4`, 32-token continuation, first-hitting + `prob_diff` confidence. Prompt:
> 
> `Once upon a time, in a small village near the sea,`
> 
> **A. PR loader + official vocab 50258 / mask 50257** (131/131 tensors loaded):
> 
> > Once upon a time, in a small village near the sea, Western Canadian University quarterback Taylor Paul played his first game in a pro varsity jersey, this time as a redshirt. Tim Woking has just revealed an unexpected
> 
> This is coherent English and typical of a small OWT LM (topic drift from fairy-tale prefix into news/sports). The port + remapped weights **do** generate, when the tokenizer matches the checkpoint.
> 
> **B. PR loader + `owt_bd3lm`’s xLM GPT2TokenizerFast** (128/131 loaded; embeddings skipped):
> 
> > Once upon a time, in a small village near the sea, vik ChefEnjoy myth Ethiopianifestyle hotlybledon demonstrate Henry credit Mostly Nationsut Moj herself Hard Fei contingency licensee accounting Slug swimming …
> 
> Unusable. Same weights, wrong vocab/mask.
> 
> ### Suggested follow-ups
> 1. Load released OWT checkpoints through `hub.repo_id` (after prefix remap, or by keeping `backbone.`), not via `+pretrained=auto` inside `__init__`.
> 2. For OWT, use a GPT-2 tokenizer whose vocab is 50258 with mask id 50257 — do not use `xlm.datamodule.GPT2TokenizerFast` as-is.
> 3. Add `Bd3lmPredictor.generate()` if demo/CLI continuation is in scope; today only `predict(batch)` exists.
> 
> Happy to turn (1)+(2) into a follow-up patch if wanted.

Fix: 1 and 2 are fixed now.

Tokenizer. Switched owt_bd3lm to load_auto_tokenizer with just a mask token, so the vocab is 50258 with the mask at 50257 and the embedding and LM head actually load now.

Loading path. Moved the transformer under self.backbone so the names match the released checkpoints, and dropped convert_hf_checkpoint.py and the init loader — hub.repo_id loads them directly now, like LLaDA.

Generation result: Gen. ppl on bd3lm-owt-block_size4 at length 1024 is 25.74 over 300 samples, against 25.70 in the paper.

Do we need to fix the 3. generate() also?

File: xlm-models/bd3lm/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+| 8 | `kuleshov-group/bd3lm-owt-block_size8` |
+| 16 | `kuleshov-group/bd3lm-owt-block_size16` |
+| 1024 | `kuleshov-group/bd3lm-owt-block_size1024-pretrain` |
+
+On a task whose vocabulary is not GPT-2's, the three vocabulary-sized tensors cannot
+transfer. They are skipped and reported, and the transformer blocks still load:
+
+```
+[bd3lm] 3 tensor(s) skipped on shape and will train from scratch - usually a vocabulary difference:
+[bd3lm]     backbone.vocab_embed.embedding: model (27, 768) vs checkpoint (50258, 768)
```
[the remark this answers is not in the document]
Comment (rjagalamari) [reply xlm-models/bd3lm/README.md]: Fixes:
Removed the star fine-tuning on released  OWT checkpoints
Renamed auto.yaml to kuleshov_group_bd3lm.yaml
